feat: cuda 11.2 and build standalone cuda images

olevski opened the pull request:
This does a few things:
1. Creates more than one version of the renkulab-py image with different versions of python (this is necessary because different versions of python are compatible with different version of tensorflow)
2. Creates cuda images that have only cuda in them (following proper version combos as imposed by tensorflow), see [here](https://www.tensorflow.org/install/source#gpu)
3. Creates the same cuda-tf images as before with different tensorflow versions that follow the official tensorflow requirements [here](https://www.tensorflow.org/install/source#gpu)
4. Updates a missing library in cuda images required by tensorflow (`libcusolver.so.11`) provided by package `libcusolver-11-2` in ubuntu

Most importantly this PR still builds the same images we were building before for renkulab-py: i.e. the one named renkulab-py:3.9-X.X.X from the same base image as before. The cuda-tf images have changed because tensorflow imposes specific requirements on the python version that cannot be maintained by always using the same base renkulab-py image with a fixed version of python.

rokroskar reviewed (changes requested):
These changes make sense to me - makes the builds a bit more complicated but I think it's also good that we cover different python versions. Could you also please update the information in the README?

olevski commented on .github/workflows/build_and_push_to_docker.yml:
```diff
@@ -69,7 +96,112 @@ jobs:
+      matrix:
+        include:
+          # taken from tensorflow compatibility chart at https://www.tensorflow.org/install/source#gpu
+          - CUDA_VERSION: "11.0"
+            BASE_IMAGE: renku/renkulab-py:python-3.8.8
+          - CUDA_VERSION: "11.2"
+            BASE_IMAGE: renku/renkulab-py:python-3.9.7
```
Ah yeah I will def do that @rokroskar.

I am sorry for dragging this out a bit. I had a really dumb mistake in my github workflow that was messing up the tests. I am pretty sure I got it now. When it is ready I will convert it from draft into a proper PR.

rokroskar replied:
no problem, thanks!

olevski commented:
@rokroskar I removed the huge table with versions and links to Dockerhub from the README and added similar lists (that are quite easier to manage IMO) in the release notes. Now the readme has a links to the releases where that large table used to be.

I think this is much better and way easier to update. Adding rows and formatting that table is really painful. And I think that such specific information clearly belongs in the release notes - not in the README.

olevski commented:
I updated the release notes for the last 4 releases with this in mind. See here: https://github.com/SwissDataScienceCenter/renkulab-docker/releases

rokroskar commented:
Thanks @olevski  - I think that's fine, we don't need to keep track of all the releases in the table. But should the readme at least have a list and description of the most recent base images? A bit like what is done at https://github.com/rocker-org/rocker with the rocker base images. The details can be left to the release notes.

rokroskar reviewed (approved):
Thanks!!!

gavin-k-lee commented:
hey @olevski - thanks for putting this together! (+ cleaning up the versions table :)) I'm just wondering why we aimed for cuda 11.2 specifically? Is this designed for limited.renku? I just checked on limited and the drivers are up to 11.5:
```
 ~ ▶ work ❯ napari-pipelines ▶ master ▶ $ ▶ nvidia-smi
Sun Jan 30 23:10:28 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 495.29.05    Driver Version: 495.29.05    CUDA Version: 11.5     |
...
```
and cuda in general is up to 11.6. Even Sv-renku is on 11.4 last time I checked (late last year), so does it make sense to build for 11.2?

olevski commented:
@gavin-k-lee I looked at the tensorflow cuda compatibility table here (https://www.tensorflow.org/install/source#gpu). And the latest version that tensorflow works with is 11.2. So I am not sure why we need to go to higher versions. I did not know that the drivers were than new. I am not sure what is the difference between these cuda packages installed in these dockerfiles and the drivers installed in the VM.

Either way I will open a new PR to add cuda version 11.6.

olevski commented:
Also looking at [here](https://pytorch.org/get-started/locally/#start-locally) for pytorch it seems that the latest cuda version pytorch supports is 11.3.
